File: OurKnowledgeBackend/src/main/java/com/example/ourknowledgebackend/service/ProjectService.java
package com.example.ourknowledgebackend.service;

import com.example.ourknowledgebackend.exceptions.DuplicateInstanceException;
import com.example.ourknowledgebackend.exceptions.InstanceNotFoundException;
import com.example.ourknowledgebackend.model.ProjectDetails;
import com.example.ourknowledgebackend.model.ProjectResult;
import com.example.ourknowledgebackend.model.entities.Participation;
import com.example.ourknowledgebackend.model.entities.Project;
import com.example.ourknowledgebackend.model.entities.Uses;
import com.example.ourknowledgebackend.model.entities.Verification;

import javax.naming.directory.InvalidAttributesException;
import java.util.List;

public interface ProjectService {

    Block<ProjectResult> listProjects(int page, int size, String keywords, Long filterId) throws InstanceNotFoundException;

    ProjectDetails projectDetails(Long id) throws InstanceNotFoundException;

    Project addProject(String name, String description, String status, String startDate, int size, List<Long> technologiesId) throws DuplicateInstanceException, InstanceNotFoundException;

    void updateProject(Long id, String name, String description, String status, String startDate, int size, Boolean updateTechnologies, List<Long> technologiesId) throws InstanceNotFoundException, DuplicateInstanceException;

    void deleteProject(Long projectId) throws InstanceNotFoundException;

    void addUses(Long projectId, Long technologyId) throws InstanceNotFoundException, DuplicateInstanceException;

    Uses deleteUses(Long usesId) throws InstanceNotFoundException;

    Participation participate(Long userId, Long projectId, String startDate, String endDate) throws InstanceNotFoundException, DuplicateInstanceException;

    Participation updateParticipate(Long participationId, String startDate, String endDate) throws InstanceNotFoundException;

    List<Verification> listVerification(Long userId, Long project) throws InstanceNotFoundException, InvalidAttributesException;

    Verification addVerification(Long userId, Long usesId) throws InstanceNotFoundException;

    Verification deleteVerification(Long verificationId, Boolean deleteKnowledge) throws InstanceNotFoundException;
}

File: our-knowledge-front/src/backend/projectService.js
import {config, appFetch} from './appFetch';

export const size = 4;

export const findProjects = (page, keywords, filterId, onSuccess) =>
    appFetch(`/project/list`,
        config('POST', {size, page, keywords, filterId}), onSuccess);

export const findByProjectId = (id, onSuccess) =>
    appFetch(`/project`,
        config('Post', {id}), onSuccess);

export const addProject = (name, description, startDate, status, size, technologyIdList, onSuccess) =>
    appFetch(`/project/add`,
        config('Post', {name, description, startDate, status, size, technologyIdList}), onSuccess);

export const updateProject = (id, name, description, startDate, status, size, updateTechnologies, technologyIdList, onSuccess) =>
    appFetch(`/project/update`,
        config('Post', {id, name, description, startDate, status, size, updateTechnologies, technologyIdList}), onSuccess);

export const addUses = (projectId, technologyId, onSuccess) =>
    appFetch(`/project/uses/add`,
        config('Post', {projectId, technologyId}), onSuccess);

export const deleteUses = (usesId, onSuccess) =>
    appFetch(`/project/uses/delete`,
        config('Post', {usesId}), onSuccess);

export const addParticipation = (userId, projectId, startDate, endDate, onSuccess) =>
    appFetch(`/project/participation/add`,
        config('Post', {userId, projectId, startDate, endDate}), onSuccess);

export const updateParticipation = (participationId, startDate, endDate, onSuccess) =>
    appFetch(`/project/participation/update`,
        config('Post', {participationId, startDate, endDate}), onSuccess);

export const addVerification = (userId, usesId, onSuccess) =>
    appFetch(`/project/verification/add`,
        config('Post', {userId, usesId}), onSuccess);

export const deleteVerification = (verificationId, deleteKnowledge, onSuccess) =>
    appFetch(`/project/verification/delete`,
        config('Post', {verificationId, deleteKnowledge}), onSuccess);


